src/libraries/System.Net.Http/src/System/Net/Http/Headers/KnownHeaders.cs

```diff
@@ -47,7 +47,7 @@ internal static class KnownHeaders
         public static readonly KnownHeader Expect = new KnownHeader("Expect", HttpHeaderType.Request | HttpHeaderType.NonTrailing, GenericHeaderParser.MultipleValueNameValueWithParametersParser, new string[] { "100-continue" }, H2StaticTable.Expect);
         public static readonly KnownHeader ExpectCT = new KnownHeader("Expect-CT");
         public static readonly KnownHeader Expires = new KnownHeader("Expires", HttpHeaderType.Content | HttpHeaderType.NonTrailing, DateHeaderParser.Parser, null, H2StaticTable.Expires);
-        public static readonly KnownHeader From = new KnownHeader("From", HttpHeaderType.Request, GenericHeaderParser.MailAddressParser, null, H2StaticTable.From);
+        public static readonly KnownHeader From = new KnownHeader("From", HttpHeaderType.Request, GenericHeaderParser.FromParser, null, H2StaticTable.From);
```

Comment:
I don't think we need a parser at all for this, even. It can be removed, like it is headers like Link or Cookie.

Reply:
I tried that at first, but the strongly typed From property on HttpRequestHeaders calls SetOrRemoveParsedValue, which asserts that there is a parser. Is there another method that From property setter can call?

Reply:
Yeah, this is a little tricky.

Strongly-typed properties currently always need some parser, due to the way that the property logic is implemented.

The generic parser that we use in other cases (e.g. Allow header) is GenericHeaderParser.TokenListParser. This does minimal validation that the value is a valid HTTP token.

However, TokenListParser allows multiple header values. Since we have a strongly-typed From property that only supports a single value, this could be problematic -- the From property logic might get confused if there are multiple parsed values.

We don't actually have the equivalent of a single-valued generic parser today. But it's easy to add this. Just as we have a definition for TokenListParser, we could add a definition for TokenParser like so:

```csharp
        internal static readonly GenericHeaderParser TokenParser = new GenericHeaderParser(false, ParseTokenList, StringComparer.OrdinalIgnoreCase);
```

The difference here is that the first constructor param, `supportsMultipleValues`, is false instead of true. You can see this same pattern with other parsers defined on GenericHeaderParser. The naming pattern is different for them, but the principle is the same.

Speaking of naming, `ParseTokenList` should really be called `ParseToken` since by itself, it only parses a token, not a list of tokens. If you want to fix that, go for it :)

Reply:
I tried this approach, but it is breaking tests. Specifically these two tests:

https://github.com/dotnet/runtime/blob/8fb2eb53f75971ce492f14bebb93ed7a236bcd6e/src/libraries/System.Net.Http/tests/UnitTests/Headers/HttpRequestHeadersTest.cs#L715-L716

https://github.com/dotnet/runtime/blob/8fb2eb53f75971ce492f14bebb93ed7a236bcd6e/src/libraries/System.Net.Http/tests/UnitTests/Headers/HttpRequestHeadersTest.cs#L728-L737

It looks like the reason is because the `ParseTokenList` method calls `HttpRuleParser.GetTokenLength`, which checks for valid "token" characters:

https://github.com/dotnet/runtime/blob/4e4b8bf2d95280e4e87881b79038d1b89f347944/src/libraries/System.Net.Http/src/System/Net/Http/HttpRuleParser.cs#L78-L84

And the ` ` (space) value is not a valid token char.

https://github.com/dotnet/runtime/blob/4e4b8bf2d95280e4e87881b79038d1b89f347944/src/libraries/System.Net.Http/src/System/Net/Http/HttpRuleParser.cs#L22-L32

So I don't believe this approach will work.

Reply:
Yeah. Yuck. Sorry for wasting your time on this.

Let's just do ParseFrom as you implemented originally.

Reply:
No worries. I just want to ensure I'm doing the right thing here.